engine/lib/upgrades/2016110900-3.0.0-denormalize_metastrings-daa6bf342e8e17dc.php

```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * Elgg 3.0.0 upgrade 2016110900
+ * denormalize_metastrings
+ *
+ * Denormalizes metastrings into metadata and annotations
+ */
+
+$db = _elgg_services()->db;
+
+$tables = [
+	'metadata',
+	'annotations',
+];
+foreach ($tables as $table) {
+	// create columns
+	$db->updateData("
+		ALTER TABLE {$db->prefix}{$table}
+		ADD COLUMN `value` text NOT NULL AFTER `entity_guid`,
+		ADD COLUMN `name` text NOT NULL AFTER `entity_guid`
+	");
+	
+	// move in all metastrings
+	$db->insertData("
+		UPDATE {$db->prefix}{$table} n_table
+		INNER JOIN {$db->prefix}metastrings msn ON n_table.name_id = msn.id
+		INNER JOIN {$db->prefix}metastrings msv ON n_table.value_id = msv.id
+		SET n_table.name = msn.string,
+		    n_table.value = msv.string
+	");
```

Comment:
This part may be quite time consuming, so it should probably use `Elgg\Upgrade\Batch`. I'm working on allowing core upgrades to also make use of it.